### PR TITLE
Add new mode to AFLplusplus

### PR DESCRIPTION
**Pull request by CarloMara**
Hello everybody,

after long discussion with @domenukk, I have completed the first patch to enable the new mode. See screenshot attached for the first general overview.

Following @domenukk advice, I've added a new flag to make this opt in, at least while it's still under test.

Cheers all,
Carlo
image

### CODE REVIEW

**Comment by domenukk** [comment]
Thank you for your contribution! The initial implementation already looks very good!
Please direct PRs towards the `dev` branch 🤌.

File: src/afl-fuzz-state.c [hunk 1]
```diff
@@ -496,6 +496,13 @@ void read_afl_environment(afl_state_t *afl, char **envp) {
             afl->max_length =
                 atoi((u8 *)get_afl_env(afl_environment_variables[i]));
 
+          }else if (!strncmp(env, "AFL_PIZZA_MODE",
```
Comment (domenukk) [inline src/afl-fuzz-state.c]: Please run `make code-format`

File: src/afl-fuzz-stats.c [hunk 2]
```diff
@@ -435,7 +435,18 @@ static void check_term_size(afl_state_t *afl) {
 /* A spiffy retro stats screen! This is called every afl->stats_update_freq
    execve() calls, plus in several other circumstances. */
 
+
 void show_stats(afl_state_t *afl) {
+  if(afl->afl_env.afl_pizza_mode){
```
Comment (domenukk) [inline src/afl-fuzz-stats.c]: This should be `likely()`

**Comment by nezza** [comment]
I can confirm that this patch works on my system as designed. Only comment would be to add TikTok support, I saw that a lot of new pizza types appear there first. How complex would that change be? And can this be extended to support pasta dishes?

**Comment by fabianfreyer** [comment]
It's lacking stats on number of cans of pineapple used, average number of pineapple pizza pieces per pizza, etc...

**Comment by andreafioraldi** [comment]
at least still 0 pineapple pizzas cooked

File: src/afl-fuzz-stats.c [hunk 3]
```diff
@@ -1219,6 +1230,788 @@ void show_stats(afl_state_t *afl) {
...
+  sprintf(tmp, "%s%s", u_stringify_int(IB(0), afl->saved_hangs),
+          (afl->saved_hangs >= KEEP_UNIQUE_HANG) ? "+" : "");
+
+  u_stringify_time_diff(time_tmp, cur_ms, afl->last_hang_time);
+  SAYF(bV bSTOP "  last conversation with customers : " cRST "%-33s     " bSTG bV bSTOP
+                "                         at table : " cRST "%-6s              " bSTG         bV "\n",
+       time_tmp, tmp);
+
+  SAYF(bVR bH bSTOP                                              cCYA
+       " Backing progress  " bSTG bH30 bH20 bH5 bX bH bSTOP cCYA
```
Comment (carpikes) [inline src/afl-fuzz-stats.c]:
```suggestion
       " Baking progress  " bSTG bH30 bH20 bH5 bX bH bSTOP cCYA 
```

**Comment by andreafioraldi** [comment]
`Testing abourted by the user` -> `Baking aborted by the chef`
second `at table` -> `# ordered peroni`

File: src/afl-fuzz-stats.c (hunk 3 above)
Comment (andreafioraldi) [reply src/afl-fuzz-stats.c]: so italian @CarloMara

File: src/afl-fuzz-state.c (hunk 1 above)
Comment (andreafioraldi) [reply src/afl-fuzz-state.c]: not even drunk kinds write `}else if` without space ziocan @CarloMara

File: src/afl-fuzz-stats.c [hunk 4]
```diff
@@ -1219,6 +1230,788 @@ void show_stats(afl_state_t *afl) {
...
+    banner_pad = (79 - banner_len) / 2;
+    memset(banner, ' ', banner_pad);
+
+#ifdef __linux__
+    if (afl->fsrv.nyx_mode) {
+
+      sprintf(banner + banner_pad,
+              "%s " cLCY VERSION cLBL " {%s} " cLGN "(%s) " cPIN "[%s] - Nyx",
+              afl->crash_mode ? cPIN "Mozzarbella Pizzeria table booking system"
+                              : cYEL "Mozzarbella Pizzeria managment system",
```
Comment (carpikes) [inline src/afl-fuzz-stats.c]:
```suggestion
                              : cYEL "Mozzarbella Pizzeria management system",
```

File: src/afl-fuzz-stats.c [hunk 5]
```diff
@@ -1219,6 +1230,788 @@ void show_stats(afl_state_t *afl) {
...
+              si, afl->use_banner, afl->power_name);
+
+    } else {
+
+#endif
+      sprintf(banner + banner_pad,
+              "%s " cLCY VERSION cLBL " {%s} " cLGN "(%s) " cPIN "[%s]",
+              afl->crash_mode ? cPIN "Mozzarbella Pizzeria table booking system"
+                              : cYEL "Mozzarbella Pizzeria managment system",
```
Comment (carpikes) [inline src/afl-fuzz-stats.c]:
```suggestion
                              : cYEL "Mozzarbella Pizzeria management system",
```

File: src/afl-fuzz-stats.c [hunk 6]
```diff
@@ -1219,6 +1230,788 @@ void show_stats(afl_state_t *afl) {
...
+       afl->sync_id ? u_stringify_int(IB(0), afl->queued_imported)
+                    : (u8 *)"n/a");
+
+  sprintf(tmp, "%s/%s, %s/%s",
+          u_stringify_int(IB(0), afl->stage_finds[STAGE_HAVOC]),
+          u_stringify_int(IB(2), afl->stage_cycles[STAGE_HAVOC]),
+          u_stringify_int(IB(3), afl->stage_finds[STAGE_SPLICE]),
+          u_stringify_int(IB(4), afl->stage_cycles[STAGE_SPLICE]));
+
+  SAYF(bV bSTOP "  18 year aniversary mode/cleaning : " cRST "%-36s  " bSTG bV bSTOP, tmp);
```
Comment (carpikes) [inline src/afl-fuzz-stats.c]:
```suggestion
  SAYF(bV bSTOP "  18 year anniversary mode/cleaning : " cRST "%-36s  " bSTG bV bSTOP, tmp);
```

File: src/afl-fuzz-stats.c [hunk 7]
```diff
@@ -1219,6 +1230,788 @@ void show_stats(afl_state_t *afl) {
...
+
+  sprintf(tmp, "%s (%s%s saved)", u_stringify_int(IB(0), afl->total_tmouts),
+          u_stringify_int(IB(1), afl->saved_tmouts),
+          (afl->saved_hangs >= KEEP_UNIQUE_HANG) ? "+" : "");
+
+  SAYF(bSTG bV bSTOP "                    burned pizzas : " cRST "%-20s" bSTG bV "\n", tmp);
+
+  /* Aaaalmost there... hold on! */
+
+  SAYF(bVR bH cCYA bSTOP " Promotional campaign on Facebook yields " bSTG bH30 bH2
```
Comment (carpikes) [inline src/afl-fuzz-stats.c]: Please don't be a boomer. 
s/Facebook/TikTok/

**Comment by fabianfreyer** [comment]
@andreafioraldi 

> `Testing abourted by the user` -> `Baking aborted by the chef` second `at table` -> `# ordered peroni`

s/chef/pizzayolo/ ?

File: src/afl-fuzz-stats.c [hunk 8]
```diff
@@ -1219,6 +1230,788 @@ void show_stats(afl_state_t *afl) {
...
+            u_stringify_int(IB(0), afl->stage_finds[STAGE_FLIP1]),
+            u_stringify_int(IB(1), afl->stage_cycles[STAGE_FLIP1]),
+            u_stringify_int(IB(2), afl->stage_finds[STAGE_FLIP2]),
+            u_stringify_int(IB(3), afl->stage_cycles[STAGE_FLIP2]),
+            u_stringify_int(IB(4), afl->stage_finds[STAGE_FLIP4]),
+            u_stringify_int(IB(5), afl->stage_cycles[STAGE_FLIP4]));
+
+  }
+
+  SAYF(bV bSTOP "                pizzas for celiac  : " cRST "%-36s  " bSTG bV bSTOP
```
Comment (carpikes) [inline src/afl-fuzz-stats.c]:
```suggestion
  SAYF(bV bSTOP "                 pizzas for celiac : " cRST "%-36s  " bSTG bV bSTOP
```

**Comment by andreafioraldi** [comment]
> s/chef/pizzayolo/ ?

A YOLO pizzaiolo?

**Comment by bitwave** [comment]
Does it support this?
grafik
If so, then please with extra banana, pineapple and kiwi.

**Comment by CarloMara** [comment]
Hello @domenukk @andreafioraldi,

I've fixed everything as suggested. IMHO it can be merged.

Cheers,
Carlo

**Pull request merged**